### PR TITLE
Fixes #140, issue not marked inactive if PR is in 'needs review state.

### LINKED ISSUE #140: Do not mark issue as inactive if referenced PR is in a "needs review" state.
If it takes 10 days to review a PR, we shouldn't bother the contributor about it. The clock for inactivity should start at the last point at which the "needs review" label is removed. 
See conversation at https://chat.zulip.org/#narrow/stream/zulipbot/topic/inactivity.20warning.

### PR DESCRIPTION
**Pull request by VibhorCodecianGupta**
Addresses #140. Attempted to fire `cleanInProgress` when either `client.cfg.activity.issues.inProgress` holds true OR the PR holds a 'needs review' label.

### CODE REVIEW

**Comment by synicalsyntax** [comment]
Hi @VibhorCodecianGupta, thanks for working on this!

I think you misunderstood @rishig's request in the issue though. He's asking in #140 that issues are not marked inactive if they have the "needs review" state. Your PR instead updates the in progress labels for an issue if the "needs review" configuration is enabled.

In order to properly resolve this issue, you would need to modify the code in `./src/automations/checkInactivity.js`.

**Comment by VibhorCodecianGupta** [comment]
Hi, I reviewed the code on your suggestions and I think that in `scrapeInactiveIssues`, right after there is an iteration over issues, before it checks whether `inactiveLabel` is present or not, a check needs to be put in place, which would fetch the reviewed/needs review status of the PR.
Only if it has been reviewed, will the code proceed to start the clock.

**Comment by synicalsyntax** [comment]
Hi @VibhorCodecianGupta, I like your reasoning, but I still don't think that's the best approach because we are checking the **referenced pull requests** for the  needs review label, not the issue itself. `scrapeInactiveIssues` is only responsible for checking the data of the issues, not any pull requests. 

A better approach would be to modify the `scrapePullRequests` function so that the `references` object carries some kind of value that lets the referenced pull request check in `scrapeInactiveIssues` know that the PR should be skipped.

**Comment by VibhorCodecianGupta** [comment]
Greetings, thanks for the valuable feedback @synicalsyntax :D 

If we could do something like `references.set(`${repoName}/${number}`, check)` where number is the PR number, and check would be a variable which would fetch whether the PR is reviews.

Then, in `scrapeInactiveIssues`, similar to what I was thinking earlier, a check could be placed when the iteration over issues begins, and the clock starts only if the PR has been reviewed, which we would know from the `${repoName}/${number}` key of the references object.
Thoughts?

**Comment by synicalsyntax** [comment]
So in `scrapePullRequests` we populate the `references` map with the key set to the repository name and issue number, and the value is the unix timestamp of when the PR was last updated. This is essential for updating the issue's last updated time according to the referenced PR, so you need to keep that. If you plan on changing the type of values stored in references, make sure to add a check comparing the type of the value.

**Comment by VibhorCodecianGupta** [comment]
Yes, I understand that. Hence I was thinking of passing an object as the value instead of `time`, and the object would contain both `time` and `number` which could be accessed

**Comment by VibhorCodecianGupta** [comment]
So it would be something like 
```
const obj = {
        timeUpdated: time,
        checkReviewed: check
      };
```
and `references.set(${repoName}/${ref}, obj);` would be pushed in `references` object.
In `scrapeInactiveIssues`, where the update time needs to accessed, `references.get(issueTag).timeUpdated` would be called. And the check where the status of the PR would be known, `references.get(issueTag).checkReviewd` would be called.
Where `check` would be `pullRequest.needsReview.label`

**Comment by synicalsyntax** [comment]
```js
const data = {
  time: 10000000, // unix time stamp 
  review: false // true if PR has reviewed label, false if not
};
```

Should be fine. Looking forward to your changes!

(I'll be offline for a few days, so thanks for your patience in getting things reviewers in advance)

**Comment by VibhorCodecianGupta** [comment]
Why should the `time` key be hardcoded? Isn't it the time when the PR is updated? 
Otherwise, I'll be sending the PR now. Thanks!

**Comment by synicalsyntax** [comment]
Yeah I just put down the value of time a single an example, definitely don't hardcore it :)

**Comment by VibhorCodecianGupta** [comment]
@synicalsyntax , The recent commits have some extra variables added to prevent the character limit exceeded test failure. Please review :D

File: src/automations/checkInactivity.js [hunk 1]
```diff
@@ -102,64 +107,72 @@ async function scrapeInactiveIssues(client, references, issues) {
   const iterator = issues[Symbol.iterator]();
 
   for (let issue of iterator) {
-    const inactiveLabel = issue.labels.find(label => {
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Why did you remove this? We still need to reject checking any issues with inactivity labels, regardless of their review status. Remove your deletions please.

File: src/automations/checkInactivity.js [hunk 2]
```diff
@@ -102,64 +107,72 @@ async function scrapeInactiveIssues(client, references, issues) {
...
-    if (time < references.get(issueTag)) time = references.get(issueTag);
+    if (references.get(issueTag).review) {
+      const inactiveLabel = issue.labels.find(label => {
+        return label.name === client.cfg.activity.inactive;
+      });
+      if (inactiveLabel) continue;
 
-    const active = client.cfg.activity.check.repositories.includes(issueTag);
+      let time = Date.parse(issue.updated_at);
+      let prUpdateTime = references.get(issueTag).time;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Should be a `const`; we don't change it. Also `prUpdateTime` isn't really camelcase?

File: src/automations/checkInactivity.js [hunk 3]
```diff
@@ -102,64 +107,72 @@ async function scrapeInactiveIssues(client, references, issues) {
...
-    if (inactiveLabel) continue;
-
-    let time = Date.parse(issue.updated_at);
     const issueNumber = issue.number;
     const repoName = issue.repository.name;
     const repoOwner = issue.repository.owner.login;
     const issueTag = `${repoName}/${issueNumber}`;
 
-    if (time < references.get(issueTag)) time = references.get(issueTag);
+    if (references.get(issueTag).review) {
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Why are you nesting everything? Just `if (!references.get(issueTag).review) continue` should be fine.

File: src/automations/checkInactivity.js (hunk 1 above)
Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: This is not deleted, it is just shifted down to a different line due to some declarations being shifted before the `needsReview` check because they needed to be declared to call `references.get(issueTag).review`

File: src/automations/checkInactivity.js (hunk 3 above)
Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: Thanks, I will immediately make the changes.

File: src/automations/checkInactivity.js (hunk 2 above)
Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: I changed `prUpdateTime` to `pullReqUpdateTime`. Also, `time` would not be a const because in the following if statement the value of `time` can be potentially updated, which is why `npm run test` returns an error.

Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: I changed `prUpdateTime` to `pullReqUpdateTime`, and made it a `const`. I'll be submitting a PR now :)

File: src/events/issues.js [hunk 4]
```diff
@@ -8,8 +8,9 @@ exports.run = (client, payload) => {
   const repoOwner = repository.owner.login;
   const repoName = repository.name;
   const l = payload.label;
+  const check = client.cfg.activity.issues.pullRequests;
 
-  if (client.cfg.activity.issues.inProgress) {
+  if (client.cfg.activity.issues.inProgress || check.needsReview.label) {
```
Comment (synicalsyntax) [inline src/events/issues.js]: No changes should be in this file.

File: src/automations/checkInactivity.js [hunk 5]
```diff
@@ -137,10 +146,12 @@ async function scrapeInactiveIssues(client, references, issues) {
       owner: repoOwner, repo: repoName, number: issueNumber, per_page: 100
     });
     const com = issueComments.data.slice(-1).pop();
+    const Body = com.body;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Why was this changed?

File: src/automations/checkInactivity.js [hunk 6]
```diff
@@ -137,10 +146,12 @@ async function scrapeInactiveIssues(client, references, issues) {
...
+    const user = com.user.login;
 
     // Use end of line comments to check if comment is from template
-    const warning = com ? com.body.endsWith("<!-- inactiveWarning -->") : null;
-    const fromClient = com ? com.user.login === client.cfg.auth.username : null;
+    const warning = com ? Body.endsWith("<!-- inactiveWarning -->") : null;
+    const fromClient = com ? user === client.cfg.auth.username : null;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Why was this changed?

File: src/automations/checkInactivity.js [hunk 7]
```diff
@@ -102,25 +107,29 @@ async function scrapeInactiveIssues(client, references, issues) {
...
 
-    if (time < references.get(issueTag)) time = references.get(issueTag);
+    if (time < pullReqUpdateTime) time = pullReqUpdateTime;
 
     const active = client.cfg.activity.check.repositories.includes(repoTag);
 
     if (time + ms >= Date.now() || !active) continue;
 
-    const aString = issue.assignees.map(assignee => assignee.login).join(", @");
+    const assignedTo = issue.assignees;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Why was this changed?

File: src/automations/checkInactivity.js [hunk 8]
```diff
@@ -102,25 +107,29 @@ async function scrapeInactiveIssues(client, references, issues) {
...
+    const issueNumber = issue.number;
+    const repoName = issue.repository.name;
+    const repoOwner = issue.repository.owner.login;
+    const issueTag = `${repoName}/${issueNumber}`;
+
     const inactiveLabel = issue.labels.find(label => {
       return label.name === client.cfg.activity.inactive;
     });
     if (inactiveLabel) continue;
+    if (!references.get(issueTag).review) continue;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: You need to check if `references.has(issueTag)` is true first before trying to fetch the issueTag from the references map. Not all issues have a reference, so you mind receive some errors like "Cannot read property review of undefined", because `references.get(issueTag)` is undefined

File: src/automations/checkInactivity.js [hunk 9]
```diff
@@ -44,10 +45,14 @@ async function scrapePullRequests(client, pullRequests) {
     }).map(c => c.commit.message);
     const bodRef = client.automations.get("issueReferenced").findKeywords(body);
 
-    if (bodRef || refIssues.length) {
-      const com = refIssues[0];
-      const ref = com ? com.match(/#([0-9]+)/)[1] : body.match(/#([0-9]+)/)[1];
-      references.set(`${repoName}/${ref}`, time);
+    if (bodRef && refIssues.length) {
+      const commitRef = refIssues[0].match(/#([0-9]+)/)[1];
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: I understand that you were trying to resolve the merge conflict, but this completely reverts the previous commit, so you need to fix this.

File: src/automations/checkInactivity.js [hunk 10]
```diff
@@ -30,6 +30,7 @@ async function scrapePullRequests(client, pullRequests) {
     const number = pullRequest.number;
     const repoName = pullRequest.base.repo.name;
     const repoOwner = pullRequest.base.repo.owner.login;
+    const check = pullRequest.reviewed.label;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Can you find a better variable name for this? Maybe `reviewedLabel`.

File: src/automations/checkInactivity.js (hunk 6 above)
Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: The above two variables were introduced so as to adhere to the less than 80 characters test case since I had earlier nested everything inside the if block, and the indentation introduced cause an overflow from the 80 character limit. Since that's been rectified, i'll rectify these changes now :)

File: src/events/issues.js (hunk 4 above)
Comment (VibhorCodecianGupta) [reply src/events/issues.js]: Yes, definitely. This change is the remnant from an earlier commit which I forgot to revert back while pushing new commits. Changing it now, thanks!

File: src/automations/checkInactivity.js (hunk 7 above)
Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: Similar reason, to successfully pass the 80 character limit test. I made a bunch of changes, seems like i couldn't revert all of them. Thanks for pointing these out :D

File: src/automations/checkInactivity.js [hunk 11]
```diff
@@ -137,7 +145,6 @@ async function scrapeInactiveIssues(client, references, issues) {
       owner: repoOwner, repo: repoName, number: issueNumber, per_page: 100
     });
     const com = issueComments.data.slice(-1).pop();
-
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Why was this line removed again?

File: src/automations/checkInactivity.js [hunk 12]
```diff
@@ -102,19 +107,22 @@ async function scrapeInactiveIssues(client, references, issues) {
...
     if (inactiveLabel) continue;
+    if (references.has(issueTag) && !references.get(issueTag).review) continue;
 
     let time = Date.parse(issue.updated_at);
-    const issueNumber = issue.number;
-    const repoName = issue.repository.name;
-    const repoOwner = issue.repository.owner.login;
-    const issueTag = `${repoName}/${issueNumber}`;
-    const repoTag = issue.repository.full_name;
+    const pullReqUpdateTime = references.get(issueTag).time;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: Maybe `pullUpdateTime` would be more appropriate?

File: src/automations/checkInactivity.js [hunk 13]
```diff
@@ -102,19 +107,22 @@ async function scrapeInactiveIssues(client, references, issues) {
...
+    const repoName = issue.repository.name;
+    const repoOwner = issue.repository.owner.login;
+    const issueTag = `${repoName}/${issueNumber}`;
+    const repoTag = issue.repository.full_name;
+
     const inactiveLabel = issue.labels.find(label => {
       return label.name === client.cfg.activity.inactive;
     });
     if (inactiveLabel) continue;
+    if (references.has(issueTag) && !references.get(issueTag).review) continue;
```
Comment (synicalsyntax) [inline src/automations/checkInactivity.js]: This statement is checking whether references has the issueTag AND whether the issueTag reference has a review property, which isn't what you want (all references already have the review property, also they aren't made in the first place). You're supposed to break out of the loop only if the reference doesn't exist.

File: src/automations/checkInactivity.js (hunk 11 above)
Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: The presence of this line was throwing a test failure 'Trailing spaces not allowed'. This was space was probably introduced in earlier commits due to declaration of new variables.

File: src/automations/checkInactivity.js (hunk 13 above)
Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: Doesn't the `references.get(issueTag).review` access `reviewedLabel`, which determines whether there is a 'reviewed' tag or not? From what I understand, the if block in question is checking whether references has the issueTag AND whether or not `reviewedLabel` is true, i.e there is a 'reviewed' label present. I think this checks whether there is a reference to the issue and if there is, then whether the PR is in 'needs review' state. If it is, then it jumps to next iteration.

If we break out of the loop only if the reference doesn't exist, then wouldn't the case where the reference exists, but the PR is in 'needs review' state be unhandled?

File: src/automations/checkInactivity.js (hunk 11 above)
Comment (synicalsyntax) [reply src/automations/checkInactivity.js]: You're probably adding trailing whitespace to this line then (it should be only a blank line with no whitespace after). Your editor might be responsible for this issue, but this line should still not be removed nonetheless.

File: src/automations/checkInactivity.js (hunk 13 above)
Comment (synicalsyntax) [reply src/automations/checkInactivity.js]: Yeah, my mistake, this is fine. Can you fix the merge conflicts in this PR and address the two other comments I made before I merge this? Thanks :)

Comment (VibhorCodecianGupta) [reply src/automations/checkInactivity.js]: The previous commit did not show any merge conflicts. Nonetheless I just pushed another commit making the changes, and hopefully it doesn't encounter a conflict

Comment (synicalsyntax) [reply src/automations/checkInactivity.js]: screenshot at dec 29 07-47-37

I still can't merge this. If you really need help with resolving Git issues, you can ask the other developers for help on [Zulip]( https://chat.zulip.org/#narrow/stream/development.20help)

**Comment by VibhorCodecianGupta** [comment]
@synicalsyntax 
#144 addresses the exact same issue with all the changes you requested. In an attempt to resolve the conflicts, the git workflow and the history became extremely disorganised and messy. Hence created a fresh branch and committed all the changes in a single commit.

You could close this PR if you'd like, #144 is the exact same :)
